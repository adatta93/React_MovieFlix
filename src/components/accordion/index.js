import React, { createContext, useContext, useState } from "react";
import {
  Body,
  Container,
  Header,
  Inner,
  Item,
  Title,
} from "./styles/accordion";

const ToggleContext = createContext();

export default function Accordion({ children, ...restProps }) {
  return <Inner {...restProps}>{children}</Inner>;
}

Accordion.Container = function AccordionContainer({ children, ...restProps }) {
  return <Container {...restProps}>{children}</Container>;
};

Accordion.Title = function AccordionTitle({ children, ...restProps }) {
  return <Title {...restProps}>{children}</Title>;
};

Accordion.Item = function AccordionItem({ children, ...restProps }) {
  const [toggleShow, setToggleShow] = useState(false);
  return (
    <ToggleContext.Provider value={{ toggleShow, setToggleShow }}>
      <Item {...restProps}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...restProps }) {
  const { toggleShow, setToggleShow } = useContext(ToggleContext);
  return (
    <Header onClick={() => setToggleShow((prevVal) => !prevVal)} {...restProps}>
      {children}
      <img
        src="/images/icons/add.png"
        alt="Open"
        className={toggleShow ? "open" : "closed"}
      />
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...restProps }) {
  const { toggleShow } = useContext(ToggleContext);
  return (
    <Body className={toggleShow ? "open" : "closed"} {...restProps}>
      <span>{children}</span>
    </Body>
  );
};
