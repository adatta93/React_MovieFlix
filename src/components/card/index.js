import React, { createContext, useContext, useState } from "react";
import {
  Container,
  Group,
  Title,
  SubTitle,
  Text,
  Item,
  Meta,
  Entities,
  Image,
  FeatureContent,
  Feature,
  FeatureTitle,
  FeatureText,
  FeatureClose,
  Maturity,
} from "./styles/card";

export const CardFeatureContext = createContext();

export default function Card({ children, ...restProps }) {
  const [showCardFeature, setShowCardFeature] = useState(false);
  const [itemFeature, setItemFeature] = useState({});
  return (
    <CardFeatureContext.Provider
      value={{
        showCardFeature,
        setShowCardFeature,
        itemFeature,
        setItemFeature,
      }}>
      <Container {...restProps}>{children}</Container>
    </CardFeatureContext.Provider>
  );
}

Card.Group = function CardGroup({ children, ...restProps }) {
  return <Group {...restProps}>{children}</Group>;
};

Card.Feature = function CardFeatuer({ category, children, ...restProps }) {
  const { showCardFeature, itemFeature, setShowCardFeature } = useContext(
    CardFeatureContext
  );
  return showCardFeature ? (
    <Feature
      {...restProps}
      src={`/images/${category}/${itemFeature.genre}/${itemFeature.slug}/large.jpg`}>
      <FeatureContent>
        <FeatureTitle>{itemFeature.title}</FeatureTitle>
        <FeatureText>{itemFeature.description}</FeatureText>
        <FeatureClose onClick={() => setShowCardFeature(false)}>
          <img src="/images/icons/close.png" alt="Close" />
        </FeatureClose>
        <Group margin="30px 0" flexDirection="row" alignItems="center">
          <Maturity rating={itemFeature.maturity}>
            {itemFeature.maturity < 12 ? "PG" : itemFeature.maturity}
          </Maturity>
          <FeatureText fontWeight="bold" textTransform="capitalize">
            {itemFeature.genre}
          </FeatureText>
        </Group>
        {children}
      </FeatureContent>
    </Feature>
  ) : null;
};

Card.Item = function CardItem({ item, children, ...restProps }) {
  const { setShowCardFeature, setItemFeature } = useContext(CardFeatureContext);
  return (
    <Item
      onClick={() => {
        setShowCardFeature(true);
        setItemFeature(item);
      }}
      {...restProps}>
      {children}
    </Item>
  );
};

Card.Entities = function CardEntities({ children, ...restProps }) {
  return <Entities {...restProps}>{children}</Entities>;
};

Card.Title = function CardTitle({ children, ...restProps }) {
  return <Title {...restProps}>{children}</Title>;
};

Card.SubTitle = function CardSubTitle({ children, ...restProps }) {
  return <SubTitle {...restProps}>{children}</SubTitle>;
};

Card.Text = function CardText({ children, ...restProps }) {
  return <Text {...restProps}>{children}</Text>;
};

Card.Meta = function CardMeta({ children, ...restProps }) {
  return <Meta {...restProps}>{children}</Meta>;
};

Card.Image = function CardImage({ ...restProps }) {
  return <Image {...restProps} />;
};
