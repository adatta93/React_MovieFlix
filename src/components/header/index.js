import React, { useState } from "react";
import { Link as ReactRouterLink } from "react-router-dom";
import {
  Background,
  ButtonLink,
  Container,
  Highlight,
  HighlightCallOut,
  Logo,
  PlayButton,
  TextLink,
  Picture,
  Group,
  Dropdown,
  Profile,
  Text,
  Search,
  SearchInput,
  SearchIcon,
} from "./styles/header";

export default function Header({ bg = true, children, ...restProps }) {
  return bg ? (
    <Background {...restProps} data-testid="header-bg">
      {children}
    </Background>
  ) : (
    children
  );
}

Header.Container = function HeaderContainer({ children, ...restProps }) {
  return <Container {...restProps}>{children}</Container>;
};

Header.Logo = function HeaderLogo({ to, ...restProps }) {
  return (
    <ReactRouterLink to={to}>
      <Logo {...restProps} />
    </ReactRouterLink>
  );
};

Header.ButtonLink = function HeaderButtonLink({ to, children, ...restProps }) {
  return (
    <ButtonLink to={to} {...restProps}>
      {children}
    </ButtonLink>
  );
};

Header.Highlight = function HeaderHighlight({ children, ...restProps }) {
  return <Highlight {...restProps}>{children}</Highlight>;
};

Header.HighlightCallOut = function HeadeHighlightCallOut({
  children,
  ...restProps
}) {
  return <HighlightCallOut {...restProps}>{children}</HighlightCallOut>;
};

Header.Text = function HeaderText({ children, ...restProps }) {
  return <Text {...restProps}>{children}</Text>;
};

Header.PlayButton = function HeaderPlayButton({ children, ...restProps }) {
  return <PlayButton {...restProps}>{children}</PlayButton>;
};

Header.Profile = function HeaderProfile({ children, ...restProps }) {
  return <Profile {...restProps}>{children}</Profile>;
};

Header.Picture = function HeaderPicture({ src, ...restProps }) {
  return <Picture src={src} {...restProps} />;
};

Header.Dropdown = function HeaderDropdown({ children, ...restProps }) {
  return <Dropdown {...restProps}>{children}</Dropdown>;
};

Header.Group = function HeaderGroup({ children, ...restProps }) {
  return <Group {...restProps}>{children}</Group>;
};

Header.TextLink = function HeaderTextLink({ children, ...restProps }) {
  return <TextLink {...restProps}>{children}</TextLink>;
};

Header.Search = function HeaderSearch({
  searchTerm,
  setSearchTerm,
  ...restProps
}) {
  const [searchActive, setSearchActive] = useState(false);

  return (
    <Search {...restProps}>
      <SearchIcon
        onClick={() => setSearchActive((prevState) => !prevState)}
        data-testid="search-click">
        <img src="/images/icons/search.png" alt="Search" />
      </SearchIcon>
      <SearchInput
        value={searchTerm}
        onChange={(e) => setSearchTerm(e.target.value)}
        placeholder="Search..."
        active={searchActive}
        data-testid="search-input"
      />
    </Search>
  );
};
