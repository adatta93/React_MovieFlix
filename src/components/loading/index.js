import React from "react";
import { LockBody, Picture, ReleaseBody, Spinner } from "./styles/loading";

export default function Loading({ src, ...restProps }) {
  return (
    <Spinner {...restProps}>
      <LockBody />
      {src ? (
        <Picture src={`/images/users/${src}.png`} data-testid="loading-pic" />
      ) : null}
    </Spinner>
  );
}

Loading.ReleaseBody = function LoadingReleaseBody() {
  return <ReleaseBody />;
};
