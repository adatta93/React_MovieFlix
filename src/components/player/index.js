import React, { createContext, useContext, useState } from "react";
import ReactDOM from "react-dom";
import { Container, Overlay, Inner, Button, VideoClose } from "./player";

export const PlayerContext = createContext();

export default function Player({ children, ...restProps }) {
  const [showPlayer, setShowPlayer] = useState(false);
  return (
    <PlayerContext.Provider value={{ showPlayer, setShowPlayer }}>
      <Container {...restProps}>{children}</Container>
    </PlayerContext.Provider>
  );
}

Player.Video = function PlayerVideo({ src, ...restProps }) {
  const { showPlayer, setShowPlayer } = useContext(PlayerContext);
  const handleClose = (e) => {
    e.stopPropagation();
    e.nativeEvent.stopImmediatePropagation();
    //console.log(e.target, e.currentTarget);
    if (e.target === e.currentTarget) {
      setShowPlayer(false);
    }
  };
  return showPlayer
    ? ReactDOM.createPortal(
        <Overlay {...restProps} onClick={handleClose} data-testid="player">
          <Inner>
            <video id="movieflix-player" controls>
              <source src={src} type="video/mp4" />
            </video>
            <VideoClose
              onClick={() => setShowPlayer(false)}
              data-testid="playerClose">
              <img src="/images/icons/close.png" alt="Close" />
            </VideoClose>
            {/* <Close /> */}
          </Inner>
        </Overlay>,
        document.body
      )
    : null;
};

Player.Button = function PlayerButton({ ...restProps }) {
  const { setShowPlayer } = useContext(PlayerContext);
  return (
    <Button {...restProps} onClick={() => setShowPlayer((prev) => !prev)}>
      Play
    </Button>
  );
};
