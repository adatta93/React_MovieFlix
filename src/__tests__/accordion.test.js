import React from "react";
import { render, fireEvent, prettyDOM } from "@testing-library/react";
import { Accordion } from "../components";
import faqData from "../fixtures/faqs.json";

describe("Accordion", () => {
  it("renders Accordion component with populated data", () => {
    const { container, getByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        {faqData.map((faq) => (
          <Accordion.Item key={faq.id}>
            <Accordion.Header>{faq.header}</Accordion.Header>
            <Accordion.Body>{faq.body}</Accordion.Body>
          </Accordion.Item>
        ))}
      </Accordion>
    );

    expect(getByText("Frequently Asked Questions")).toBeTruthy();
    expect(getByText("What is MovieFlix?")).toBeTruthy();
    expect(getByText("How much does MovieFlix cost?")).toBeTruthy();
    expect(getByText("Where can I watch?")).toBeTruthy();
    expect(getByText("How do I cancel?")).toBeTruthy();
    expect(getByText("What can I watch on MovieFlix?")).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
