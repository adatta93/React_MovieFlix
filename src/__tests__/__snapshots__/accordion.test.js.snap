// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders Accordion component with populated data 1`] = `
<div
  class="accordion__Inner-sc-1wyxntn-1 dMjIbL"
>
  <h1
    class="accordion__Title-sc-1wyxntn-2 bQmdFZ"
  >
    Frequently Asked Questions
  </h1>
  <div
    class="accordion__Item-sc-1wyxntn-3 eaiwW"
  >
    <div
      class="accordion__Header-sc-1wyxntn-4 bZXLcX"
    >
      What is MovieFlix?
      <img
        alt="Open"
        class="closed"
        src="/images/icons/add.png"
      />
    </div>
    <div
      class="accordion__Body-sc-1wyxntn-5 jFnqOY closed"
    >
      <span>
        MovieFlix is a streaming service that offers a wide variety of award-winning TV programmes, films, anime, documentaries and more – on thousands of internet-connected devices.

You can watch as much as you want, whenever you want, without a single advert – all for one low monthly price. There's always something new to discover, and new TV programmes and films are added every week!
      </span>
    </div>
  </div>
  <div
    class="accordion__Item-sc-1wyxntn-3 eaiwW"
  >
    <div
      class="accordion__Header-sc-1wyxntn-4 bZXLcX"
    >
      How much does MovieFlix cost?
      <img
        alt="Open"
        class="closed"
        src="/images/icons/add.png"
      />
    </div>
    <div
      class="accordion__Body-sc-1wyxntn-5 jFnqOY closed"
    >
      <span>
        Watch MovieFlix on your smartphone, tablet, smart TV, laptop or streaming device, all for one low fixed monthly fee. Plans start from £5.99 a month. No extra costs or contracts.
      </span>
    </div>
  </div>
  <div
    class="accordion__Item-sc-1wyxntn-3 eaiwW"
  >
    <div
      class="accordion__Header-sc-1wyxntn-4 bZXLcX"
    >
      Where can I watch?
      <img
        alt="Open"
        class="closed"
        src="/images/icons/add.png"
      />
    </div>
    <div
      class="accordion__Body-sc-1wyxntn-5 jFnqOY closed"
    >
      <span>
        Watch anywhere, anytime, on an unlimited number of devices. Sign in with your MovieFlix account to watch instantly on the web at MovieFlix.com from your personal computer or on any internet-connected device that offers the MovieFlix app, including smart TVs, smartphones, tablets, streaming media players and game consoles.

You can also download your favourite programmes with the iOS, Android, or Windows 10 app. Use downloads to watch while you're on the go and without an internet connection. Take MovieFlix with you anywhere.
      </span>
    </div>
  </div>
  <div
    class="accordion__Item-sc-1wyxntn-3 eaiwW"
  >
    <div
      class="accordion__Header-sc-1wyxntn-4 bZXLcX"
    >
      How do I cancel?
      <img
        alt="Open"
        class="closed"
        src="/images/icons/add.png"
      />
    </div>
    <div
      class="accordion__Body-sc-1wyxntn-5 jFnqOY closed"
    >
      <span>
        MovieFlix is flexible. There are no annoying contracts and no commitments. You can easily cancel your account online in two clicks. There are no cancellation fees – start or stop your account at any time.
      </span>
    </div>
  </div>
  <div
    class="accordion__Item-sc-1wyxntn-3 eaiwW"
  >
    <div
      class="accordion__Header-sc-1wyxntn-4 bZXLcX"
    >
      What can I watch on MovieFlix?
      <img
        alt="Open"
        class="closed"
        src="/images/icons/add.png"
      />
    </div>
    <div
      class="accordion__Body-sc-1wyxntn-5 jFnqOY closed"
    >
      <span>
        MovieFlix has an extensive library of feature films, documentaries, TV programmes, anime, award-winning MovieFlix originals, and more. Watch as much as you want, any time you want.
      </span>
    </div>
  </div>
</div>
`;
